File: algorithms/01_python/numerical/lu_decomposition.py
"""LU decomposition with partial pivoting (Numerical Recipes 2.3)."""
def lu_solve(matrix: list[list[float]], rhs: list[float]) -> list[float]:
    n = len(matrix)
    a = [row[:] for row in matrix]
    b = rhs[:]
    for col in range(n):
        pivot = max(range(col, n), key=lambda r: abs(a[r][col]))
        a[col], a[pivot] = a[pivot], a[col]
        b[col], b[pivot] = b[pivot], b[col]
        for row in range(col + 1, n):
            factor = a[row][col] / a[col][col]
            for k in range(col, n):
                a[row][k] -= factor * a[col][k]
            b[row] -= factor * b[col]
    x = [0.0] * n
    for row in range(n - 1, -1, -1):
        total = b[row] - sum(a[row][k] * x[k] for k in range(row + 1, n))
        x[row] = total / a[row][row]
    return x


if __name__ == "__main__":
    x = lu_solve([[2, 1, -1], [-3, -1, 2], [-2, 1, 2]], [8, -11, -3])
    assert all(abs(a - b) < 1e-9 for a, b in zip(x, [2, 3, -1]))
    print("lu decomposition ok")

File: algorithms/03_cpp/numerical/lu_decomposition.cpp
// LU decomposition with partial pivoting (Numerical Recipes 2.3).
#include <algorithm>
#include <cassert>
#include <cmath>
#include <iostream>
#include <vector>

std::vector<double> luSolve(std::vector<std::vector<double>> a, std::vector<double> b) {
    int n = static_cast<int>(a.size());
    for (int col = 0; col < n; ++col) {
        int pivot = col;
        for (int r = col + 1; r < n; ++r)
            if (std::abs(a[r][col]) > std::abs(a[pivot][col])) pivot = r;
        std::swap(a[col], a[pivot]);
        std::swap(b[col], b[pivot]);
        for (int r = col + 1; r < n; ++r) {
            double f = a[r][col] / a[col][col];
            for (int k = col; k < n; ++k) a[r][k] -= f * a[col][k];
            b[r] -= f * b[col];
        }
    }
    std::vector<double> x(n);
    for (int r = n - 1; r >= 0; --r) {
        double s = b[r];
        for (int k = r + 1; k < n; ++k) s -= a[r][k] * x[k];
        x[r] = s / a[r][r];
    }
    return x;
}

int main() {
    auto x = luSolve({{2, 1, -1}, {-3, -1, 2}, {-2, 1, 2}}, {8, -11, -3});
    assert(std::abs(x[0] - 2) < 1e-9 && std::abs(x[1] - 3) < 1e-9 && std::abs(x[2] + 1) < 1e-9);
    std::cout << "lu decomposition ok\n";
    return 0;
}
